### ISSUE TITLE
Socket connections opened after executing connect.Disconnect(service_instance)

### ISSUE THREAD

**Issue by inevu** [vmware/pyvmomi#194]
Hi All,

It seems like some file descriptor sockets remain opened after executing connect.Disconnect(service_instance). I am running an infinite loop in this example. 

This is part of the code:

```
while True:
    service_instance = connect.SmartConnect(host=hostname, user=username, pwd=password, port=int(port))
    content = service_instance.RetrieveContent()
    perf_dict = {}
    perfList = content.perfManager.perfCounter
    for counter in perfList:
        counter_full = "{}.{}.{}".format(counter.groupInfo.key, counter.nameInfo.key, counter.rollupType)
        perf_dict[counter_full] = counter.key
    connect.Disconnect(service_instance)
    time.sleep(20)
```
If I go to /proc/PID/fd/ I see this:

```
lrwx------ 1 user group 64 Dec  3 00:06 0 -> /dev/pts/7
lrwx------ 1 user group 64 Dec  3 00:06 1 -> /dev/pts/7
lrwx------ 1 user group 64 Dec  3 00:31 10 -> socket:[7646375]
lrwx------ 1 user group 64 Dec  3 00:52 11 -> socket:[7646684]
lrwx------ 1 user group 64 Dec  3 00:06 12 -> /tmp/ffiu4HSiE (deleted)
lrwx------ 1 user group 64 Dec  3 00:52 13 -> socket:[7647465]
lrwx------ 1 user group 64 Dec  3 00:52 14 -> socket:[7647162]
lrwx------ 1 user group 64 Dec  3 00:22 15 -> socket:[7644543]
lrwx------ 1 user group 64 Dec  3 00:05 2 -> /dev/pts/7
lrwx------ 1 user group 64 Dec  3 00:06 3 -> socket:[7645020]
lrwx------ 1 user group 64 Dec  3 00:06 4 -> socket:[7644691]
lrwx------ 1 user group 64 Dec  3 00:06 5 -> socket:[7645646]
lrwx------ 1 user group 64 Dec  3 00:30 6 -> socket:[7645164]
lrwx------ 1 user group 64 Dec  3 00:30 7 -> socket:[7645601]
lrwx------ 1 user group 64 Dec  3 00:30 8 -> socket:[7647017]
lrwx------ 1 user group 64 Dec  3 00:30 9 -> socket:[7646093]
```

I thought the Disconnect() method would clear all these sockets, but it doesn't. Is this expected? Am I doing something wrong?

Thanks!

**Comment by frispete** [comment]
@inevu: please tell us your python version: specifically, Python 2.6 or later?

**Comment by parthgala** [comment]
@inevu Which Python version are you using? There is a bug in Python 2.6 or earlier where in sockets aren't getting closed properly (http://bugs.python.org/issue10127)

**Comment by nlamirault** [comment]
Any status for this bug ?

**Comment by nlamirault** [comment]
Same behavior using Python 3.4

**Comment by hartsock** [comment]
I'm setting aside time for this bug and the SSL bug in the next 2 weeks.

**Comment by netmason** [comment]
May I inquire whether there are updates on this bug?   Thanks.

**Comment by nlamirault** [comment]
Any news ?

**Comment by shyamachilles** [comment]
how do I get around this problem?

**Comment by hartsock** [comment]
cc @tianhao64

**Comment by shyamachilles** [comment]
Is this the same issue as I see in https://github.com/vmware/pyvmomi/issues/294

**Comment by svdgraaf** [comment]
Any news? I'm experiencing the same issue. App running in Gunicorn (if that matters at all?), python 2.7.9, Debian. If you need more details or test data, please let me know.

**Comment by sreejithkesavan** [comment]
We too are experiencing the same issue. Any updates on when this will get fixed?

**Comment by hartsock** [comment]
I'm sorry, I'm not even on this project anymore. I'll send this over to the current maintainer.

Attn: @tianhao64

**Comment by svdgraaf** [comment]
Hey @tianhao64 any news? Would love to get in touch and help out. We're a paying Vmware customer, we could also file an issue internally with you guys if you point us in the right direction :)

**Comment by tianhao64** [comment]
Hey @svdgraaf , I think this is already fixed in 6.0 release by (https://github.com/vmware/pyvmomi/commit/1574c9eafb9960a6654ec658721ee0aa0726ec3b)

I tested using both 2.7 and 3.4 and can't reproduce it:
```
Python 2.7.10 (default, Oct 23 2015, 18:05:06) 
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.59.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from ssl import _create_unverified_context
>>> from pyVim import connect
>>> connection = connect.SmartConnect(host='10.160.13.64', user='root', pwd='ca$hc0w', sslContext=_create_unverified_context())
/Users/het/virenv/279pyvmomi/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py:791: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
  InsecureRequestWarning)
>>> connection.RetrieveContent().rootFolder.childEntity
(ManagedObject) [
   'vim.Datacenter:ha-datacenter'
]
>>> connect.Disconnect(connection)
>>> connection.RetrieveContent().rootFolder.childEntity
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/het/virenv/279pyvmomi/lib/python2.7/site-packages/pyVmomi/VmomiSupport.py", line 564, in __call__
    return self.f(*args, **kwargs)
  File "/Users/het/virenv/279pyvmomi/lib/python2.7/site-packages/pyVmomi/VmomiSupport.py", line 384, in _InvokeAccessor
    return self._stub.InvokeAccessor(self, info)
  File "/Users/het/virenv/279pyvmomi/lib/python2.7/site-packages/pyVmomi/StubAdapterAccessorImpl.py", line 47, in InvokeAccessor
    raise objectContent.missingSet[0].fault
pyVmomi.VmomiSupport.NotAuthenticated: (vim.fault.NotAuthenticated) {
   dynamicType = <unset>,
   dynamicProperty = (vmodl.DynamicProperty) [],
   msg = '',
   faultCause = <unset>,
   faultMessage = (vmodl.LocalizableMessage) [],
   object = 'vim.Folder:ha-folder-root',
   privilegeId = 'System.View'
}
>>> connection = connect.SmartConnect(host='10.160.13.64', user='root', pwd='ca$hc0w', sslContext=_create_unverified_context())
/Users/het/virenv/279pyvmomi/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py:791: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
  InsecureRequestWarning)
>>> connection.RetrieveContent().rootFolder.childEntity
(ManagedObject) [
   'vim.Datacenter:ha-datacenter'
]
```

```
Python 3.4.3 (v3.4.3:9b73f1c3e601, Feb 23 2015, 02:52:03) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from ssl import _create_unverified_context                                                                                                                                               >>> from pyVim import connect
>>> connection = connect.SmartConnect(host='xxx', user='xxx', pwd='xxx', sslContext=_create_unverified_context())
/Users/het/virenv/343pyvmomi/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py:791: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
  InsecureRequestWarning)
>>> connection.RetrieveContent().rootFolder.childEntity
(ManagedObject) [
   'vim.Datacenter:ha-datacenter'
]
>>> connect.Disconnect(connection)
>>> connection.RetrieveContent().rootFolder.childEntity
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/het/virenv/343pyvmomi/lib/python3.4/site-packages/pyVmomi/VmomiSupport.py", line 564, in __call__
    return self.f(*args, **kwargs)
  File "/Users/het/virenv/343pyvmomi/lib/python3.4/site-packages/pyVmomi/VmomiSupport.py", line 384, in _InvokeAccessor
    return self._stub.InvokeAccessor(self, info)
  File "/Users/het/virenv/343pyvmomi/lib/python3.4/site-packages/pyVmomi/StubAdapterAccessorImpl.py", line 47, in InvokeAccessor
    raise objectContent.missingSet[0].fault
pyVmomi.VmomiSupport.NotAuthenticated: (vim.fault.NotAuthenticated) {
   dynamicType = <unset>,
   dynamicProperty = (vmodl.DynamicProperty) [],
   msg = '',
   faultCause = <unset>,
   faultMessage = (vmodl.LocalizableMessage) [],
   object = 'vim.Folder:ha-folder-root',
   privilegeId = 'System.View'
}
>>> connection = connect.SmartConnect(host='xxx', user='xxx', pwd='xxx', sslContext=_create_unverified_context())
/Users/het/virenv/343pyvmomi/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py:791: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
  InsecureRequestWarning)
>>> connection.RetrieveContent().rootFolder.childEntity
(ManagedObject) [
   'vim.Datacenter:ha-datacenter'
]
```

**Comment by tianhao64** [comment]
Forgot to mark this as fixed after the release. My apologies. Please let me know if you still experiencing issues with new code.

**Comment by svdgraaf** [comment]
Interestingly, I'm still having the timeout in 2.7.9. I'm checking if it is fixed by playing with the timeout from #347

**Comment by svdgraaf** [comment]
I can confirm that raising the timeout fixes the issue
